triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala

```diff
@@ -87,84 +89,124 @@ object Server {
+
+  final case class GetServerBinding(replyTo: ActorRef[ServerBinding]) extends Message
+  final case object Stop extends Message
+
+  private final case class StartFailed(cause: Throwable) extends Message
+  private final case class Started(binding: ServerBinding) extends Message
+
+  final case class TriggerStarting(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message])
+      extends Message
+  final case class TriggerStarted(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message])
+      extends Message
+  final case class TriggerInitializationFailure(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message],
+      cause: String
+  ) extends Message
+  final case class TriggerRuntimeFailure(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message],
+      cause: String
+  ) extends Message
 
   def apply(
       host: String,
       port: Int,
       ledgerConfig: LedgerConfig,
       maxInboundMessageSize: Int,
+      maxFailureNumberOfRetries: Int,
+      failureRetryTimeRange: Duration,
       dar: Option[Dar[(PackageId, Package)]],
   ): Behavior[Message] = Behaviors.setup { ctx =>
+    val server = new Server(dar)
+
     // http doesn't know about akka typed so provide untyped system
     implicit val untypedSystem: akka.actor.ActorSystem = ctx.system.toClassic
     implicit val materializer: Materializer = Materializer(untypedSystem)
     implicit val esf: ExecutionSequencerFactory =
       new AkkaExecutionSequencerPool("TriggerService")(untypedSystem)
 
-    dar.foreach(addDar(compiledPackages, _))
+    def startTrigger(
+        ctx: ActorContext[Server.Message],
+        token: (Jwt, JwtPayload),
+        trigger: Trigger,
+        params: StartParams,
+        ledgerConfig: LedgerConfig,
+        maxInboundMessageSize: Int,
+        maxFailureNumberOfRetries: Int,
+        failureRetryTimeRange: Duration,
+    ): JsValue = {
+      val jwt: Jwt = token._1
+      val jwtPayload: JwtPayload = token._2
+      val party: Party = Party(jwtPayload.party);
+      val uuid = UUID.randomUUID
+      val ident = uuid.toString
+      val ref = ctx.spawn(
+        TriggerRunner(
+          new TriggerRunner.Config(
+            ctx.self,
+            uuid,
+            jwt,
+            server.compiledPackages,
+            trigger,
+            ledgerConfig,
+            maxInboundMessageSize,
+            maxFailureNumberOfRetries,
+            failureRetryTimeRange,
+            party),
+          ident
+        ),
+        ident + "-monitor"
+      )
+      JsObject(("triggerId", uuid.toString.toJson))
+    }
+
+    def stopTrigger(uuid: UUID, token: (Jwt, JwtPayload)): JsValue = {
+      //TODO(SF, 2020-05-20): At least check that the provided token
+      //is the same as the one used to start the trigger and fail with
+      //'Unauthorized' if not (expect we'll be able to do better than
```

Comment:
Do you have a plan for how to do better than this?

Reply:
As a matter of a fact I think I do. My idea is that if presented with a token that doesn't match the start token, if the old and new token agree in party, the old token has expired and (in the absence of a better API) we can use the new token to get an ACS (or establish the conditions that starting a trigger would entail), we might conclude we are indeed looking at a valid stop.

Reply:
This idea has the nice property that it does not require the user presenting a token that has expired. Anyway, what do you think of that?